lib/elixir/lib/module/types/descr.ex

```diff
@@ -202,4 +232,91 @@ defmodule Module.Types.Descr do
         (mask &&& val) !== 0,
         do: {type, [], []}
   end
+
+  defp bitmap_count_top_types(val), do: Integer.digits(val, 2) |> Enum.sum()
+
+  ## Atoms
+
+  # The atom component of a type consists of pairs `{MapSet.t(atom()), boolean()}`.
+  # If the boolean is true, the type represents the union of the atoms in the set.
+  # If the boolean is false, the type represents all the possible atoms, except the ones
+  # in the set.
+  #
+  # Example:
+  #   - `{MapSet.new([:a, :b]), true}` represents type `:a or :b`
+  #   - `{MapSet.new([:c, :d]), false}` represents type `atom() \ (:c or :d)
+  #
+  # `{MapSet.new([]), false}` is the `atom()` top type, as it is the difference
+  # of `atom()` with an empty list.
+
+  defp atom_mk(a) when is_atom(a), do: {MapSet.new([a]), true}
+  defp atom_mk(as) when is_list(as), do: {MapSet.new(as), true}
+
+  defp atom_intersection({s1, neg1}, {s2, neg2}) do
+    result_set =
+      case {neg1, neg2} do
+        {true, true} -> MapSet.intersection(s1, s2)
+        {false, false} -> MapSet.union(s1, s2)
+        {true, false} -> MapSet.difference(s1, s2)
+        {false, true} -> MapSet.difference(s2, s1)
+      end
+
+    if MapSet.size(result_set) == 0 do
+      0
+    else
+      {result_set, neg1 or neg2}
+    end
+  end
+
+  defp atom_union({s1, true}, {s2, true}), do: {MapSet.union(s1, s2), true}
+  defp atom_union({s1, false}, {s2, false}), do: {MapSet.intersection(s1, s2), false}
+  defp atom_union({s1, true}, {s2, false}), do: {MapSet.difference(s2, s1), false}
+  defp atom_union({s1, false}, {s2, true}), do: {MapSet.difference(s1, s2), false}
+
+  defp atom_difference({s1, neg1}, {s2, neg2}) do
+    result_set =
+      case {neg1, neg2} do
+        {true, true} -> MapSet.difference(s1, s2)
+        {false, false} -> MapSet.difference(s2, s1)
+        {true, false} -> MapSet.intersection(s1, s2)
+        {false, true} -> MapSet.union(s1, s2)
+      end
+
+    if MapSet.size(result_set) == 0 do
+      0
+    else
+      {result_set, neg1 or not neg2}
+    end
+  end
+
+  defp literal(lit), do: {:__block__, [], [lit]}
+  @boolset MapSet.new([true, false])
+
+  defp atom_to_quoted({a = %MapSet{}, true}) do
+    if MapSet.subset?(@boolset, a) do
+      MapSet.difference(a, @boolset)
+      |> Enum.sort()
+      |> Enum.map(&literal/1)
+      |> Enum.reduce({:boolean, [], []}, &{:or, [], [&2, &1]})
```

Comment:
Sadly this doesn't work because if we have `[:a, :b]` to be made into a union, we need to apply `literal` to each one of them (for printing); but with this code the reduce will take the first one as initial value of the accumulator e.g. we get `{:or, [], [:a, {:__block__, [], :b}]}`.